app/models/work_packages/scopes/allowed_to.rb

```diff
@@ -0,0 +1,184 @@
+# -- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2010-2023 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+
+module WorkPackages::Scopes
+  module AllowedTo
+    extend ActiveSupport::Concern
+
+    class_methods do
+      # Returns an ActiveRecord::Relation to find all work packages for which
+      # +user+ has the given +permission+ either directly on the work package
+      # or by the linked project
+      def allowed_to(user, permission) # rubocop:disable Metrics/AbcSize
+        permissions = Authorization.contextual_permissions(permission, :work_package, raise_on_unknown: true)
+
+        if user.admin? && permissions.all?(&:grant_to_admin?)
+          where(id: allowed_to_admin_relation(permissions))
+        elsif user.anonymous?
+          where(id: allowed_to_non_member_relation(user, permissions))
+        else
+          union = Arel::Nodes::UnionAll.new(
+            Arel::Nodes::UnionAll.new(
+              allowed_to_member_relation(user, permissions).select(arel_table[:id]).arel,
```

Comment:
There is a very subtle bug in this part of the statement. Abbreviated the statement looks like this:

```sql
SELECT "work_packages".*
FROM "work_packages"
WHERE "work_packages"."id" IN ((
	(SELECT "work_packages"."id"
		FROM "members"
		INNER JOIN "projects" ON "projects"."active" = TRUE
		AND "members"."project_id" = "work_packages"."project_id"
		AND "members"."user_id" = 5
		AND "members"."entity_type" = 'WorkPackage'
		INNER JOIN "work_packages" ON "members"."entity_id" = "work_packages"."id"
		INNER JOIN "enabled_modules" ON "projects"."id" = "enabled_modules"."project_id"
		AND "enabled_modules"."name" IN ('work_package_tracking')
		AND "projects"."active" = TRUE
		INNER JOIN "member_roles" ON "member_roles"."member_id" = "members"."id"
		INNER JOIN "roles" ON "roles"."id" = "member_roles"."role_id"
		INNER JOIN "member_roles" "member_roles_members" ON "member_roles_members"."member_id" = "members"."id"
		INNER JOIN "role_permissions" ON "roles"."id" = "role_permissions"."role_id"
		AND (FALSE
							OR "role_permissions"."permission" = 'view_work_packages'
							AND "enabled_modules"."name" = 'work_package_tracking'))
UNION ALL
  // work packages in public project
UNION ALL
  // work packages in public project or in projects in which the user has the necessary permission via a role
))
```

Or further reduced to the interesting part

```sql
SELECT "work_packages".*
FROM "work_packages"
WHERE "work_packages"."id" IN ((
	(SELECT "work_packages"."id"
		FROM "members"
		INNER JOIN "projects" ON "projects"."active" = TRUE
		AND "members"."project_id" = "work_packages"."project_id"
		// some more conditions
		INNER JOIN "work_packages" ON "members"."entity_id" = "work_packages"."id"
		// some more joins
UNION ALL
  // work packages in public project
UNION ALL
  // work packages in public project or in projects in which the user has the necessary permission via a role
))
```

Please notice that the `AND "members"."project_id" = "work_packages"."project_id"` condition is before the `INNER JOIN "work_packages"` in that first part of the UNIONed id collection. 

Because the `work_packages` table hasn't been joined in the subquery yet, PostgreSQL will reference the work packages table in the outer statement  
```sql
SELECT "work_packages".*
FROM "work_packages"
```
and do ... something. It is just horribly slow on a larger dataset so I never managed to find out if the results are correct in the end.

Swapping the order fixes the problem (and performs way quicker). 

Swapping would lead to the following statement

```sql
(SELECT "work_packages"."id"
		FROM "members"
	    INNER JOIN "work_packages" ON "members"."entity_id" = "work_packages"."id" AND "members"."entity_type" = 'WorkPackage'
		INNER JOIN "projects" ON "projects"."active" = TRUE
		AND "members"."project_id" = "work_packages"."project_id"
		AND "members"."user_id" = 1445
		AND "members"."entity_type" = 'WorkPackage'		
		INNER JOIN "enabled_modules" ON "projects"."id" = "enabled_modules"."project_id"
		AND "enabled_modules"."name" IN ('work_package_tracking')
		AND "projects"."active" = TRUE
		INNER JOIN "member_roles" ON "member_roles"."member_id" = "members"."id"
		INNER JOIN "roles" ON "roles"."id" = "member_roles"."role_id"
		INNER JOIN "member_roles" "member_roles_members" ON "member_roles_members"."member_id" = "members"."id"
		INNER JOIN "role_permissions" ON "roles"."id" = "role_permissions"."role_id"
		AND (FALSE
							OR "role_permissions"."permission" = 'view_work_packages'
							AND "enabled_modules"."name" = 'work_package_tracking'))
```

Given the `project_id` column on the `members` table this can be simplified. There is also an unnecessary second join with member_roles in here `INNER JOIN "member_roles" "member_roles_members" ON "member_roles_members"."member_id" = "members"."id"` That can be removed. The result could look like this:

```sql
(SELECT "members".entity_id
    FROM members	    
		INNER JOIN "projects" ON "projects"."active" = TRUE
		AND "members"."project_id" = "projects"."id"
		AND "members"."user_id" = 1445
		AND "members"."entity_type" = 'WorkPackage'		
		INNER JOIN "enabled_modules" ON "projects"."id" = "enabled_modules"."project_id"
		AND "enabled_modules"."name" IN ('work_package_tracking')
		AND "projects"."active" = TRUE
		INNER JOIN "member_roles" ON "member_roles"."member_id" = "members"."id"
		INNER JOIN "roles" ON "roles"."id" = "member_roles"."role_id"
		INNER JOIN "role_permissions" ON "roles"."id" = "role_permissions"."role_id"
		AND (FALSE
							OR "role_permissions"."permission" = 'view_work_packages'
							AND "enabled_modules"."name" = 'work_package_tracking'))
```							

That unnecessary second join with member_roles `INNER JOIN "member_roles" "member_roles_members" ON "member_roles_members"."member_id" = "members"."id"` seems to have crept in from the `Project.allowed_to` scope implementation and can be removed there as well.

I don't know though whether removing the join with the `work_packages` table out of the statement is worth that much. Performance-wise, it did not make a difference in my tests. So if the statement for fetching the allowed WorkPackages is the only (or the most important) use case for populating the `project_id` column for work package members, then we should rethink whether this is worth the effort of maintaining the column (e.g. when moving a work package between projects).